File: scripts/util/camera.py
#!/usr/bin/env python3
# Copyright 2004-present Facebook. All Rights Reserved.

# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

import json
import math

import numpy as np
import numpy.polynomial.polynomial as poly
from logger import Logger
from matrix_operations import is_unitary, normalize_vector
from ray import Ray
from scipy import linalg
from scipy.spatial.transform import Rotation as R

K_NEAR_INFINITY = 1e4
INFINITY = 1e50

log = Logger(__name__)
log.logger.propagate = False


class Camera:

    """The Camera class contains all attributes of a camera such as:
        type of lens, rotations, resolution, principal, focal, etc.

    Attributes:
        id (string) : name of camera
        type (string) : type of camera lens i.e. ftheta, rectilinear, orthographic, equiosolid
        group (string) : name of group to which the camera belongs

        position (numpy 1x3 array) : position of the camera relative to center of the rig (meters)
        rotation (numpy 3x3 "matrix"/2D array) : matrix that represents the rotation of the camera
            right
            up
            backward

        resolution (numpy 1x2 array) : resolution of camera in pixels
        focal (numpy 1x2 array) : number of pixels per radian at principal point
        principal (numpy 1x2 dim array) : pixel coordinate of the optical axis

        distortion (numpy 1x3 array) : array of coefficients for the distortion polynomial
        cos_fov (float) : cosine value of the fov in radians
    """

    def __init__(self, json_file=None, json_string=None):
        if json_file:
            self.load_from_json_file(json_file)
        elif json_string:
            self.load_from_json_string(json_string)

    def load_from_json_file(self, json_file):
        with open(json_file) as f:
            json_string = f.read()
        self.load_from_json_string(json_string)

    def load_from_json_string(self, json_string):
        json_obj = json.loads(json_string)
        log.check_ge(float(json_obj["version"]), 1.0)
        self.version = float(json_obj["version"])
        self.id = str(json_obj["id"])
        self.type = str(json_obj["type"])
        if "group" in json_obj:
            self.group = str(json_obj["group"])

        self.position = deserialize_list(json_obj["origin"], 3)

        self.set_rotation(json_obj["forward"], json_obj["up"], json_obj["right"])

        self.resolution = deserialize_list(json_obj["resolution"], 2)

        self.focal = deserialize_list(json_obj["focal"], 2)

        if "principal" in json_obj:
            self.principal = deserialize_list(json_obj["principal"], 2)
        else:
            self.principal = self.resolution / 2

        if "distortion" in json_obj:
            self.set_distortion(deserialize_list(json_obj["distortion"]))
        else:
            self.set_default_distortion()

        if "fov" in json_obj:
            self.set_fov(float(json_obj["fov"]))
        else:
            self.set_default_fov()

    def serialize(self):
        json_obj = {
            "version": self.version,
            "id": self.id,
            "type": self.type,
            "origin": self.position.tolist(),
            "forward": self.forward().tolist(),
            "up": self.up().tolist(),
            "right": self.right().tolist(),
            "resolution": self.resolution.tolist(),
            "focal": self.focal.tolist(),
        }

        if hasattr(self, "group"):
            json_obj["group"] = self.group
        if not np.array_equal(self.principal, self.resolution / 2):
            json_obj["principal"] = self.principal.tolist()
        if np.any(self.__distortion):
            json_obj["distortion"] = self.__distortion.tolist()
        if not self.is_default_fov():
            json_obj["fov"] = self.get_fov()

        return json_obj

    """ Orientation """

    def forward(self):
        return np.negative(self.rotation[2])

    def backward(self):
        return self.rotation[2]

    def up(self):
        return self.rotation[1]

    def right(self):
        return self.rotation[0]

    def get_rotation(self):
        return self.rotation

    def set_rotation(self, forward, up, right):
        # TODO: might have to convert to angle axis and back to rotation matrix if there is
        #      a variance from floating point error
        log.check_lt(
            np.dot(np.cross(right, up), forward), 0, "rotation must be right-handed"
        )
        self.rotation = np.array([right, up, np.negative(forward)])

        tol = 0.001
        log.check(is_unitary(self.rotation, tol))

    def get_rotation_vector(self):
        r = R.from_dcm(self.get_rotation())
        return r.as_rotvec()

    def set_rotation_vector(self, rotvec):
        r = R.from_rotvec(rotvec)
        self.rotation = r.as_dcm()

    def get_scalar_focal(self):
        log.check_eq(self.focal[0], -self.focal[1], "pixels must be square")
        return self.focal[0]

    def set_scalar_focal(self, scalar):
        self.focal = np.asarray([scalar, -scalar])

    """ FOV """

    def get_fov(self):
        return math.acos(self.cos_fov)

    def set_fov(self, fov):
        self.cos_fov = math.cos(fov)
        log.check_ge(self.cos_fov, get_default_cos_fov(self.type))

    def set_default_fov(self):
        self.cos_fov = get_default_cos_fov(self.type)

    def is_default_fov(self):
        return self.cos_fov == get_default_cos_fov(self.type)

    def is_behind(self, point):
        return np.dot(self.backward(), point - self.position) >= 0

    def is_outside_fov(self, point):
        if self.cos_fov == -1:
            return False
        if self.cos_fov == 0:
            return self.is_behind(point)
        v = point - self.position
        dotprod = np.dot(self.forward(), v)
        v_norm_squared = math.pow(linalg.norm(v), 2)
        return (
            dotprod * abs(dotprod) <= self.cos_fov * abs(self.cos_fov) * v_norm_squared
        )

    def is_outside_image_circle(self, pixel):
        if self.is_default_fov():
            return False

        # find an edge point by projecting a point from the FOV cone
        sin_fov = math.sqrt(1 - self.cos_fov * self.cos_fov)
        edge = self.camera_to_sensor(np.array([0, sin_fov, -self.cos_fov]))

        # pixel is outside FOV if its farther from the principal than the edge point
        sensor = (pixel - self.principal) / self.focal
        return math.pow(linalg.norm(sensor), 2) >= linalg.norm(edge)

    def is_outside_sensor(self, pixel):
        pixX = pixel[0]
        pixY = pixel[1]
        resX = self.resolution[0]
        resY = self.resolution[1]
        return not (0 <= pixX and pixX < resX and 0 <= pixY and pixY < resY)

    def sees(self, point):
        if self.is_outside_fov(point):
            return (False, None)
        pixel = self.world_to_pixel(point)
        return (not self.is_outside_sensor(pixel), pixel)

    def overlap(self, camera):
        """estimate the fraction of the frame that is covered by the other camera"""
        # just brute force probeCount x probeCount points
        kProbeCount = 10
        inside = 0
        for y in range(0, kProbeCount):
            for x in range(0, kProbeCount):
                p = np.array([x, y]) * self.resolution / (kProbeCount - 1)
                if (
                    not self.is_outside_image_circle(p)
                    and camera.sees(self.point_near_infinity(p))[0]
                ):
                    inside += 1

        return inside / math.pow(kProbeCount, 2)

    """ Rescale """

    def rescale(self, resolution):
        self.principal = self.principal * (np.asarray(resolution) / self.resolution)
        self.focal = self.focal * (np.asarray(resolution) / self.resolution)
        self.resolution = np.asarray(resolution)

    def normalize(self):
        self.principal = self.principal / self.resolution
        self.focal = self.focal / self.resolution
        self.resolution = np.ones(2)

    def is_normalized(self):
        return np.array_equal(self.resolution, np.ones(2))

    """ Distortion """

    def get_distortion(self):
        return self.__distortion

    def get_distortion_max(self):
        return self.__distortion_max

    def set_default_distortion(self):
        self.__distortion = np.zeros(3)
        self.__distortion_max = INFINITY

    def set_distortion(self, distortion):
        count = distortion.size
        while distortion[count - 1] == 0:
            count -= 1
            if count == 0:
                return self.set_default_distortion()

        # distortion polynomial is x + d[0] * x^3 + d[1] * x^5 ...
        # derivative is: 1 + d[0] * 3 x^2 + d[1] * 5 x^4 ...
        # using y = x^2: 1 + d[0] * 3 y + d[1] * 5 y^2 ...
        derivative = np.empty(count + 1)
        derivative[0] = 1
        for i in range(0, count):
            derivative[i + 1] = distortion[i] * (2 * i + 3)

        # find smallest rea; root greater than zero in the derivative
        roots = poly.polyroots(derivative)
        y = INFINITY
        for root in roots:
            if np.isreal(root) and 0 < root and root < y:
                y = np.real(root)  # "convert" to real due to casting warnings

        self.__distortion = distortion
        self.__distortion_max = math.sqrt(y)

    # distortion is modeled in pixel space as:
    #   distort(r) = r + d0 * r^3 + d1 * r^5
    def distort(self, r):
        r = min(r, self.__distortion_max)
        return self.__distort_factor(r * r) * r

    def __distort_factor(self, r_squared):
        index = self.get_distortion().size - 1
        result = self.get_distortion()[index]
        while index > 0:
            index -= 1
            result = self.get_distortion()[index] + r_squared * result
        return 1 + r_squared * result

    def undistort(self, y):
        if np.count_nonzero(self.get_distortion()) == 0:
            return y
        if y >= self.distort(self.__distortion_max):
            return self.__distortion_max

        smidgen = 1.0 / K_NEAR_INFINITY
        k_max_steps = 10

        x0 = 0
        y0 = 0
        dy0 = 1
        for _step in range(0, k_max_steps):
            x1 = (y - y0) / dy0 + x0
            y1 = self.distort(x1)
            if abs(y1 - y) < smidgen:
                return x1
            dy1 = (self.distort(x1 + smidgen) - y1) / smidgen
            log.check_ge(dy1, 0, "went past a maximum")
            x0 = x1
            y0 = y1
            dy0 = dy1
        return x0  # should not happen

    """ projection """

    def world_to_pixel(self, point):
        """Compute pixel coordinates from point in rig space"""
        # transform point in rig space to camera space
        camera_point = np.matmul(self.rotation, (np.asarray(point) - self.position))
        # transform point in camera space to distorted sensor coordinates
        sensor_point = self.camera_to_sensor(camera_point)
        # transform distorted sensor coordinates to pixel coordinates
        return sensor_point * self.focal + self.principal

    def pixel_to_world(self, pixel, depth=None):
        """Compute rig coordinates from pixel and return as a parameterized line"""
        # transform from pixel to distorted sensor coordinates
        sensor_point = (pixel - self.principal) / self.focal
        # transform from distorted sensor coordinates to unit camera vector
        unit = self.sensor_to_camera(sensor_point)
        # transform from camera space to rig space
        ray = Ray(self.position, np.matmul(self.rotation.T, unit))
        if depth:
            return ray.point_at(depth)
        else:
            return ray

    def camera_to_sensor(self, camera_point):
        # FTHETA: r = theta
        # RECTILINEAR: r = tan(theta)
        # EQUISOLID: r = 2 sin(theta / 2)
        # ORTHOGRAPHIC: r = sin(theta)
        # see https://wiki.panotools.org/Fisheye_Projection
        if self.type == "FTHETA":
            # r = theta <=>
            # r = atan2(|xy|, -z)
            xy = linalg.norm(camera_point[:2])
            unit = normalize_vector(camera_point[:2])
            r = math.atan2(xy, -camera_point[2])
            return self.distort(r) * unit
        elif self.type == "RECTILINEAR":
            # r = tan(theta) <=>
            # r = |xy| / -z <=>
            # pre-distortion result is xy / -z
            xy = linalg.norm(camera_point[:2])
            unit = normalize_vector(camera_point[:2])
            if -camera_point[2] <= 0:  # outside FOV
                r = math.tan(math.pi / 2)
            else:
                r = xy / -camera_point[2]
            return self.distort(r) * unit
        elif self.type == "EQUISOLID":
            # r = 2 sin(theta / 2) <=>
            # using sin(theta / 2) = sqrt((1 - cos(theta)) / 2)
            # r = 2 sqrt((1 + z / |xyz|) / 2)
            xy = linalg.norm(camera_point[:2])
            unit = normalize_vector(camera_point[:2])
            r = 2 * math.sqrt((1 + camera_point[2] / linalg.norm(camera_point)) / 2)
            return self.distort(r) * unit
        elif self.type == "ORTHOGRAPHIC":
            # r = sin(theta) <=>
            # r = |xy| / |xyz| <=>
            # pre-distortion result is xy / |xyz|
            xy = linalg.norm(camera_point[:2])
            unit = normalize_vector(camera_point[:2])
            if -camera_point[2] <= 0:  # outside FOV
                r = math.sin(math.pi / 2)
            else:
                r = linalg.norm(camera_point[:2] / linalg.norm(camera_point))
            return self.distort(r) * unit
        else:
            log.fatal(f"unexpected type {self.type}")

    def sensor_to_camera(self, sensor_point):
        norm = linalg.norm(sensor_point)
        norm_squared = math.pow(norm, 2)

        if norm_squared == 0:
            return np.array([0, 0, -1])

        r = self.undistort(norm)
        if self.type == "FTHETA":
            theta = r
        elif self.type == "RECTILINEAR":
            theta = math.atan(r)
        elif self.type == "EQUISOLID":
            # arcsin function is undefined outside the interval [-1, 1]
            theta = 2 * math.asin(r / 2) if r <= 2 else math.pi
        elif self.type == "ORTHOGRAPHIC":
            theta = math.asin(r) if r <= 1 else math.pi / 2
        else:
            log.fatal(f"unexpected type {self.type}")

        unit = np.asarray(math.sin(theta) / norm * sensor_point)
        unit = np.append(unit, -math.cos(theta))
        return unit

    def point_near_infinity(self, pixel):
        return self.pixel_to_world(pixel, K_NEAR_INFINITY)


def get_default_cos_fov(type):
    if type == "RECTILINEAR" or type == "ORTHOGRAPHIC":
        return 0  # hemisphere
    elif type == "FTHETA" or type == "EQUISOLID":
        return -1  # sphere
    else:
        log.fatal(f"unexpected type {type}")


def deserialize_list(list, length=None):
    if length is not None:
        log.check_eq(len(list), length, "list is not the expected length")
    return np.asarray(list)

File: source/util/Camera.cpp
/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "source/util/Camera.h"

#include <unsupported/Eigen/Polynomials>

#include <folly/Format.h>

namespace fb360_dep {

const Camera::Real Camera::kNearInfinity = 1e4;

Camera::Camera(const Type type, const Vector2& res, const Vector2& focal)
    : type(type), resolution(res), focal(focal) {
  position.setZero();
  rotation.setIdentity();
  principal = resolution / 2;
  setDefaultDistortion();
  setDefaultFov();
}

Camera::Camera(const folly::dynamic& json) {
  CHECK_GE(json["version"].asDouble(), 1.0);

  id = json["id"].getString();

  type = deserializeType(json["type"]);

  position = deserializeVector<3>(json["origin"]);

  setRotation(
      deserializeVector<3>(json["forward"]),
      deserializeVector<3>(json["up"]),
      deserializeVector<3>(json["right"]));

  resolution = deserializeVector<2>(json["resolution"]);

  if (json.count("principal")) {
    principal = deserializeVector<2>(json["principal"]);
  } else {
    principal = resolution / 2;
  }

  if (json.count("distortion")) {
    const folly::dynamic& entry = json["distortion"];
    CHECK_LE(entry.size(), getDistortion().size()) << "bad distortion " << entry;
    Distortion distortion;
    for (int i = 0; i < distortion.size(); ++i) {
      distortion[i] = i < int(entry.size()) ? entry[i].asDouble() : 0;
    }
    setDistortion(distortion);
  } else {
    setDefaultDistortion();
  }

  if (json.count("fov")) {
    setFov(json["fov"].asDouble());
  } else {
    setDefaultFov();
  }

  focal = deserializeVector<2>(json["focal"]);

  if (json.count("group")) {
    group = json["group"].getString();
  }
}

void Camera::setRotation(const Vector3& forward, const Vector3& up, const Vector3& right) {
  CHECK_LT(right.cross(up).dot(forward), 0) << "rotation must be right-handed";
  rotation.row(2) = -forward; // +z is back
  rotation.row(1) = up; // +y is up
  rotation.row(0) = right; // +x is right
  // re-unitarize
  const Camera::Real tol = 0.001;
  CHECK(rotation.isUnitary(tol)) << rotation << " is not close to unitary";
  Eigen::AngleAxis<Camera::Real> aa(rotation);
  rotation = aa.toRotationMatrix();
}

void Camera::setRotation(const Vector3& forward, const Vector3& up) {
  setRotation(forward, up, forward.cross(up));
}

void Camera::setRotation(const Vector3& angleAxis) {
  // convert angle * axis to rotation matrix
  Real angle = angleAxis.norm();
  Vector3 axis = angleAxis / angle;
  if (angle == 0) {
    axis = Vector3::UnitX();
  }
  rotation = Eigen::AngleAxis<Real>(angle, axis).toRotationMatrix();
}

Camera::Vector3 Camera::getRotation() const {
  // convert rotation matrix to angle * axis
  Eigen::AngleAxis<Real> angleAxis(rotation);
  if (angleAxis.angle() > M_PI) {
    angleAxis.angle() = 2 * M_PI - angleAxis.angle();
    angleAxis.axis() = -angleAxis.axis();
  }

  return angleAxis.angle() * angleAxis.axis();
}

void Camera::setDefaultDistortion() {
  distortion_.setZero();
  distortionMax_ = INFINITY;
}

void Camera::setDistortion(const Distortion& distortion) {
  // ignore trailing zeros
  Eigen::Index count = distortion.size();
  while (distortion[count - 1] == 0) {
    if (--count == 0) {
      return setDefaultDistortion();
    }
  }

  // distortion polynomial is x + d[0] * x^3 + d[1] * x^5 ...
  // derivative is: 1 + d[0] * 3 x^2 + d[1] * 5 x^4 ...
  // using y = x^2: 1 + d[0] * 3 y + d[1] * 5 y^2 ...
  VLOG(2) << "Solving for camera distortions..."; // HACK: Prevents seg fault in solver
  Eigen::Matrix<Real, Eigen::Dynamic, 1> derivative(count + 1);
  derivative[0] = 1;
  for (int i = 0; i < count; ++i) {
    derivative[i + 1] = distortion[i] * (2 * i + 3);
  }

  // find real roots in derivative
  Eigen::PolynomialSolver<Real, Eigen::Dynamic> solver;
  solver.compute(derivative);
  std::vector<Real> roots;
  solver.realRoots(roots);

  // find smallest root greater than zero
  Real y = INFINITY;
  for (const Real& root : roots) {
    if (0 < root && root < y) {
      y = root;
    }
  }

  distortion_ = distortion;
  distortionMax_ = sqrt(y);
}

#ifndef SUPPRESS_RIG_IO

folly::dynamic Camera::serialize() const {
  // clang-format off
  folly::dynamic result = folly::dynamic::object("version", 1)("type", serializeType(type))(
      "origin", serializeVector(position))("forward", serializeVector(forward()))(
      "up", serializeVector(up()))("right", serializeVector(right()))(
      "resolution", serializeVector(resolution))("focal", serializeVector(focal))("id", id);
  // clang-format on
  if (principal != resolution / 2) {
    result["principal"] = serializeVector(principal);
  }
  if (!getDistortion().isZero()) {
    result["distortion"] = serializeVector(getDistortion());
  }
  if (!isDefaultFov()) {
    result["fov"] = getFov();
  }
  if (!group.empty()) {
    result["group"] = group;
  }

  return result;
}

#endif // SUPPRESS_RIG_IO

void Camera::setScalarFocal(const Real& scalar) {
  focal = {scalar, -scalar};
}

Camera::Real Camera::getScalarFocal() const {
  CHECK_EQ(focal.x(), -focal.y()) << "pixels are not square";
  return focal.x();
}

Camera::Real Camera::getDefaultCosFov(Camera::Type type) {
  switch (type) {
    case Camera::Type::RECTILINEAR:
    case Camera::Type::ORTHOGRAPHIC:
      return 0; // hemisphere
    default:
      return -1; // sphere
  }
}

void Camera::setDefaultFov() {
  cosFov = getDefaultCosFov(type);
}

void Camera::setFov(const Real& fov) {
  cosFov = std::cos(fov);
  CHECK(cosFov >= getDefaultCosFov(type));
}

Camera::Real Camera::getFov() const {
  return std::acos(cosFov);
}

bool Camera::isDefaultFov() const {
  return cosFov == getDefaultCosFov(type);
}

Camera Camera::rescale(const Vector2& newResolution) const {
  Camera result = *this;
  result.principal.array() *= newResolution.array() / result.resolution.array();
  result.focal.array() *= newResolution.array() / result.resolution.array();
  result.resolution = newResolution;
  return result;
}

void Camera::normalize() {
  principal = principal.cwiseQuotient(resolution);
  focal = focal.cwiseQuotient(resolution);
  resolution = Vector2::Ones();
}

bool Camera::isNormalized() const {
  return resolution == Vector2::Ones();
}

// WARNING: modifies input cameras
void Camera::normalizeRig(Camera::Rig& rig) {
  for (Camera& cam : rig) {
    if (!cam.isNormalized()) {
      cam.normalize();
    }
  }
}

Camera::Rig Camera::loadRig(const filesystem::path& filename) {
  std::string json;
  folly::readFile(filename.string().c_str(), json);
  CHECK(!json.empty()) << "could not read JSON file: " << filename;
  return loadRigFromJsonString(json);
}

Camera::Rig Camera::loadRigFromJsonString(const std::string& json) {
  folly::dynamic dynamic = folly::parseJson(json);
  std::vector<Camera> cameras;
  for (const auto& camera : dynamic["cameras"]) {
    cameras.emplace_back(camera);
  }
  return cameras;
}

void Camera::perturbCameras(
    std::vector<Camera>& cameras,
    const double posAmount,
    const double rotAmount,
    const double principalAmount,
    const double focalAmount) {
  for (auto& camera : cameras) {
    if (&camera != &cameras[0]) {
      perturb(camera.position, posAmount);
      auto rotation = camera.getRotation();
      perturb(rotation, rotAmount);
      camera.setRotation(rotation);
    }
    perturb(camera.principal, principalAmount);
    if (focalAmount != 0) {
      Camera::Real scalarFocal = camera.getScalarFocal();
      perturbScalar(scalarFocal, focalAmount);
      camera.setScalarFocal(scalarFocal);
    }
  }
}

const Camera& Camera::findCameraById(const std::string id, const Camera::Rig& rig) {
  for (const Camera& camera : rig) {
    if (camera.id == id) {
      return camera;
    }
  }
  LOG(FATAL) << folly::sformat("Camera id {} not found", id);
}

#ifndef SUPPRESS_RIG_IO

void Camera::saveRig(
    const std::string& filename,
    const std::vector<Camera>& cameras,
    const std::vector<std::string>& comments,
    const int doubleNumDigits) {
  folly::dynamic dynamic = folly::dynamic::object("cameras", folly::dynamic::array());
  for (const auto& camera : cameras) {
    dynamic["cameras"].push_back(camera.serialize());
  }
  if (!comments.empty()) {
    dynamic["comments"] = folly::dynamic::array(comments.begin(), comments.end());
  }
  folly::json::serialization_opts opts;
  opts.sort_keys = true;
  opts.pretty_formatting = true;
  if (doubleNumDigits > 0) {
    opts.double_mode = double_conversion::DoubleToStringConverter::FIXED;
    opts.double_num_digits = doubleNumDigits;
  }
  folly::writeFile(folly::json::serialize(dynamic, opts), filename.c_str());
}

#endif // SUPPRESS_RIG_IO

} // namespace fb360_dep
